crates/sui-core/src/consensus_adapter.rs

```diff
@@ -361,60 +523,43 @@ impl ConsensusAdapter {
-                    .submit_to_consensus(&transaction, epoch_store)
-                    .await
-                {
-                    // This can happen during Narwhal reconfig, so wait for a few retries.
-                    if retries > 3 {
-                        error!(
-                            "Error submitting transaction to own narwhal worker: {:?}",
-                            e
-                        );
-                    }
-                    self.opt_metrics.as_ref().map(|metrics| {
-                        metrics.sequencing_certificate_failures.inc();
-                    });
-                    retries += 1;
-                    time::sleep(Duration::from_secs(10)).await;
-                }
+        let should_submit = self.should_submit(
+            epoch_store.committee(),
+            &self.authority,
+            &transaction,
+            exclusions.clone(),
+        );
 
-                // we want to record the num of retries when reporting latency but to avoid label
-                // cardinality we do some simple bucketing to give us a good enough idea of how
-                // many retries happened associated with the latency.
-                let bucket = match retries {
-                    0..=10 => retries.to_string(), // just report the retry count as is
-                    11..=20 => "between_10_and_20".to_string(),
-                    21..=50 => "between_20_and_50".to_string(),
-                    51..=100 => "between_50_and_100".to_string(),
-                    _ => "over_100".to_string(),
-                };
+        if should_submit {
+            self.submit_after_selected(transaction.clone(), epoch_store)
+                .await;
+        }
 
+        // We wait to ensure the transaction was received by consensus. If sufficient time passes
+        // and consensus hasn't received the transaction, we submit it ourselves as a fallback.
+        let sleep_timer = tokio::time::sleep(Duration::from_secs(7));
+        let processed_waiter = epoch_store
+            .consensus_message_processed_notify(transaction.key())
+            .boxed();
+        match select(processed_waiter, Box::pin(sleep_timer)).await {
+            Either::Left((processed, _)) => {
+                processed.expect("Storage error when waiting for consensus message processed");
+            }
+            Either::Right(((), _timer)) => {
+                // we timed out
+                // try again with exclusion of the validator that was selected to submit
                 self.opt_metrics.as_ref().map(|metrics| {
-                    metrics
-                        .sequencing_acknowledge_latency
-                        .with_label_values(&[&bucket])
-                        .observe(ack_start.elapsed().as_secs_f64());
+                    metrics.sequencing_certificate_timeouts.inc();
                 });
+                self.submit_after_selected(transaction.clone(), epoch_store)
+                    .await;
```

Comment:
I would like to note that after a second look, the original code has a max sleep time of 10 seconds, and since we add 7s seconds as a multiplier to each position, if we had (and with updates to this PR) currently have a latency of 10 seconds or more, we will submit from every authority. I've changed it to remain how it is, but its worth keeping in mind as we evaluate going forward.

Reply:
@laura-makdah could you please point the code where we do have the max sleep time of 10 seconds - just want to make sure we are talking about the same thing. When looking to the original `submit_delay_certificate` method, I can see that the maximum delay multiplier is 10, so the maximum backoff time will be 10*7 = 70 seconds , meaning that any validator after the position 10 will have waited for maximum 70 seconds, and then yes, all of them would go and submit the transaction.

Reply:
Ah, you're right